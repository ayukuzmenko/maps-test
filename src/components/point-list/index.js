export { default } from './point-list';
