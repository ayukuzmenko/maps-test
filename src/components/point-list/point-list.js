import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';
import { reoderPoint } from '../../actions';
import Point from '../point';

class PointList extends Component {
  onDragEnd = result => {
    if (!result.destination) {
      return;
    }

    const { points, reoderPoint } = this.props;

    const OrderedPoints = reorder(points, result.source.index, result.destination.index);

    reoderPoint(OrderedPoints);
  };

  get pointBody() {
    const { points } = this.props;

    return points.map((item, index) => (
      <Draggable key={item.id} draggableId={item.id} index={index}>
        {(provided, snapshot) => (
          <div ref={provided.innerRef} {...provided.draggableProps} {...provided.dragHandleProps}>
            <Point text={item.text} />
          </div>
        )}
      </Draggable>
    ));
  }

  render() {
    const { points } = this.props;

    if (!points.length) {
      return null;
    }

    return (
      <DragDropContext onDragEnd={this.onDragEnd}>
        <Droppable droppableId="droppable">
          {(provided, snapshot) => (
            <div ref={provided.innerRef} {...provided.droppableProps}>
              {this.pointBody}
              {provided.placeholder}
            </div>
          )}
        </Droppable>
      </DragDropContext>
    );
  }
}

const reorder = (list, startIndex, endIndex) => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};

const mapStateToProps = state => {
  return {
    points: state.points,
  };
};

export default connect(
  mapStateToProps,
  { reoderPoint },
)(PointList);
