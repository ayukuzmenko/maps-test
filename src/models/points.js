export const points = {
  state: [],
  reducers: {
    reoderPoints: (state, payload) => (state = payload),

    addNewPoint: (state, payload) => {
      state.push({
        id: payload.newPointId,
        coords: payload.coords,
        adress: payload.adress,
        loaded: true,
      });
      return state;
    },

    deletePoint: (state, payload) => {
      return state.filter(item => item.id !== payload);
    },

    replacePoint: (state, payload) => {
      state[payload.arrIndex] = {
        id: state[payload.arrIndex].id,
        coords: payload.coords,
        adress: payload.adress,
        loaded: true,
      };
    },
  },
  effects: dispatch => ({
    async searchPoint(payload) {
      window.ymaps.geocode(payload, { result: 1 }).then(res => {
        const geoObj = res.geoObjects.get(0);
        if (geoObj) {
          this.addNewPoint({
            adress: geoObj.getAddressLine(),
            coords: geoObj.geometry.getCoordinates(),
            newPointId: (Date.now() + Math.random()).toString(),
          });
        } else {
          // if result not found
        }
      });
      // .catch(error => {
      //   dispatch({
      //     type: NEW_POINT + FAIL,
      //     error,
      //   });
      // });
    },

    async getNewCoords(payload) {
      window.ymaps.geocode(payload.newCoords, { result: 1 }).then(res => {
        const geoObj = res.geoObjects.get(0);
        if (res) {
          const adress = geoObj.getAddressLine() ? geoObj.getAddressLine() : `Address unknown`;
          dispatch.mPoints.replacePoint({
            adress: adress,
            coords: payload.newCoords,
            arrIndex: payload.arrIndex,
          });
        } else {
          // if result not found
        }
      });
    },
  }),
};
