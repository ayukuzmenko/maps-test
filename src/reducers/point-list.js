import { REORDER_POINTS } from '../constants';

export default (points = [`test1`, `test2`, `test3`, `test4`], action) => {
  if (action.type === REORDER_POINTS) {
    return (points = action.payload.pointList);
  } else {
    return points;
  }
};
