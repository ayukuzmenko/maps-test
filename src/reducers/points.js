import { REORDER_POINTS, NEW_POINT, DELETE_POINT, SUCCESS } from '../constants';
import { Record, List } from 'immutable';

const pointRecord = new Record({
  id: null,
  text: null,
  loading: false,
  loaded: false,
  error: null,
});

export default (points = List([]), action) => {
  const { type, payload } = action;
  switch (type) {
    case REORDER_POINTS: {
      return (points = payload.pointList);
    }
    case NEW_POINT + SUCCESS: {
      return points.push(
        new pointRecord({
          id: action.newPointId,
          text: payload.text,
          coords: payload.coords,
          loaded: true,
        }),
      );
    }
    case DELETE_POINT: {
      return points.filter(item => item.id !== payload.id);
    }
    default: {
      return points;
    }
  }
};
