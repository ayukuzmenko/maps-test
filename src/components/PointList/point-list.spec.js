import React from 'react';
import { mount } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';
import { PointList } from './point-list';
import { init } from '@rematch/core';
import points from '../../models/points';
import { Provider } from 'react-redux';

const store = init({
  models: { points },
});

describe('Serch bar component test', () => {
  it('Should match to snapshot', () => {
    const props = {
      points: [
        {
          adress: 'Россия, Москва',
          coords: [55.753215, 37.622504],
          id: '1554114179293.4595',
        },
      ],
    };

    const searchBarContainer = mount(
      <Provider store={store}>
        <PointList {...props} />
      </Provider>,
    );

    expect(mountToJson(searchBarContainer)).toMatchSnapshot();
  });
});
