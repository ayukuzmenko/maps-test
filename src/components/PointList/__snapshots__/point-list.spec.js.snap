// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serch bar component test Should match to snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "model": [Function],
      "name": "0",
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <PointList
    points={
      Array [
        Object {
          "adress": "Россия, Москва",
          "coords": Array [
            55.753215,
            37.622504,
          ],
          "id": "1554114179293.4595",
        },
      ]
    }
  >
    <DragDropContext
      onDragEnd={[Function]}
    >
      <Connect(Droppable)
        direction="vertical"
        droppableId="droppable"
        ignoreContainerClipping={false}
        isCombineEnabled={false}
        isDropDisabled={false}
        type="DEFAULT"
      >
        <Droppable
          direction="vertical"
          draggingFromThisWith={null}
          draggingOverWith={null}
          droppableId="droppable"
          ignoreContainerClipping={false}
          isCombineEnabled={false}
          isDraggingOver={false}
          isDropDisabled={false}
          placeholder={null}
          shouldAnimatePlaceholder={true}
          type="DEFAULT"
          updateViewportMaxScroll={[Function]}
        >
          <DroppableDimensionPublisher
            direction="vertical"
            droppableId="droppable"
            getDroppableRef={[Function]}
            getPlaceholderRef={[Function]}
            ignoreContainerClipping={false}
            isCombineEnabled={false}
            isDropDisabled={false}
            type="DEFAULT"
          >
            <div
              data-react-beautiful-dnd-droppable="0"
            >
              <Connect(Draggable)
                disableInteractiveElementBlocking={false}
                draggableId="1554114179293.4595"
                index={0}
                isDragDisabled={false}
                key="1554114179293.4595"
                shouldRespectForceTouch={true}
              >
                <Draggable
                  disableInteractiveElementBlocking={false}
                  draggableId="1554114179293.4595"
                  dragging={null}
                  drop={[Function]}
                  dropAnimationFinished={[Function]}
                  index={0}
                  isDragDisabled={false}
                  lift={[Function]}
                  move={[Function]}
                  moveByWindowScroll={[Function]}
                  moveDown={[Function]}
                  moveLeft={[Function]}
                  moveRight={[Function]}
                  moveUp={[Function]}
                  secondary={
                    Object {
                      "combineTargetFor": null,
                      "offset": Object {
                        "x": 0,
                        "y": 0,
                      },
                      "shouldAnimateDisplacement": true,
                    }
                  }
                  shouldRespectForceTouch={true}
                >
                  <DraggableDimensionPublisher
                    draggableId="1554114179293.4595"
                    droppableId="droppable"
                    getDraggableRef={[Function]}
                    index={0}
                    key="1554114179293.4595"
                    type="DEFAULT"
                  >
                    <DragHandle
                      callbacks={
                        Object {
                          "onCancel": [Function],
                          "onDrop": [Function],
                          "onLift": [Function],
                          "onMove": [Function],
                          "onMoveDown": [Function],
                          "onMoveLeft": [Function],
                          "onMoveRight": [Function],
                          "onMoveUp": [Function],
                          "onWindowScroll": [Function],
                        }
                      }
                      canDragInteractiveElements={false}
                      draggableId="1554114179293.4595"
                      getDraggableRef={[Function]}
                      getShouldRespectForceTouch={[Function]}
                      isDragging={false}
                      isDropAnimating={false}
                      isEnabled={true}
                    >
                      <div
                        aria-roledescription="Draggable item. Press space bar to lift"
                        data-react-beautiful-dnd-drag-handle="0"
                        data-react-beautiful-dnd-draggable="0"
                        draggable={false}
                        onBlur={[Function]}
                        onDragStart={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onMouseDown={[Function]}
                        onTouchStart={[Function]}
                        onTransitionEnd={null}
                        style={
                          Object {
                            "transform": null,
                            "transition": null,
                          }
                        }
                        tabIndex={0}
                      >
                        <Connect(Point)
                          item={
                            Object {
                              "adress": "Россия, Москва",
                              "coords": Array [
                                55.753215,
                                37.622504,
                              ],
                              "id": "1554114179293.4595",
                            }
                          }
                        >
                          <Point
                            deletePoint={[Function]}
                            item={
                              Object {
                                "adress": "Россия, Москва",
                                "coords": Array [
                                  55.753215,
                                  37.622504,
                                ],
                                "id": "1554114179293.4595",
                              }
                            }
                          >
                            <div>
                              <span
                                className="test"
                              >
                                Россия, Москва
                              </span>
                              <button
                                onClick={[Function]}
                              >
                                x
                              </button>
                            </div>
                          </Point>
                        </Connect(Point)>
                      </div>
                    </DragHandle>
                  </DraggableDimensionPublisher>
                </Draggable>
              </Connect(Draggable)>
              <AnimateInOut
                on={null}
                shouldAnimate={true}
              />
            </div>
          </DroppableDimensionPublisher>
        </Droppable>
      </Connect(Droppable)>
    </DragDropContext>
  </PointList>
</Provider>
`;
